sdks/java/io/hadoop/input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -562,29 +563,53 @@ private void validateUserInputForKeyAndValue() throws IOException, InterruptedEx
      * parameterized type's type arguments of InputFormat. If parameterized type has any type
      * arguments such as T, K, V, etc then validation is done by encoding and decoding key/value
      * object of first pair returned by RecordReader.
+     * @throws InterruptedException
+     * @throws IOException
      */
     private <T> boolean validateClass(String inputFormatGenericClassName, Coder coder,
-        Object object, String property) {
+        String property) throws IOException, InterruptedException {
+      RecordReader<?, ?> reader = null;
       try {
         Class<?> inputClass = Class.forName(inputFormatGenericClassName);
         /*
          * Validates key/value class with InputFormat's parameterized type.
          */
-        if (property.equals("key.class")) {
-          return (conf.getHadoopConfiguration().getClass("key.class",
-              Object.class)).isAssignableFrom(inputClass);
+        if (property.equals("key.class")
+            && !(conf.getHadoopConfiguration().getClass("key.class", Object.class))
+                .isAssignableFrom(inputClass)) {
+          this.setExpectedKeyClass(inputClass);
+          return false;
+        }
+        if (property.equals("value.class")
+            && !(conf.getHadoopConfiguration().getClass("value.class", Object.class))
+                .isAssignableFrom(inputClass)) {
+          this.setExpectedValueClass(inputClass);
+          return false;
         }
-        return (conf.getHadoopConfiguration().getClass("value.class",
-            Object.class)).isAssignableFrom(inputClass);
       } catch (ClassNotFoundException e) {
         /*
          * Given inputFormatGenericClassName is a type parameter i.e. T, K, V, etc. In such cases
-         * class validation for user provided input key/value will not work correctly. Therefore
-         * the need to validate key/value class by encoding and decoding key/value object with
-         * the given coder.
+         * class validation for user provided input key/value will not work correctly. Therefore the
+         * need to validate key/value class by encoding and decoding key/value object with the given
+         * coder.
          */
-        return checkEncodingAndDecoding((Coder<T>) coder, (T) object);
+        reader = fetchFirstRecordReader();
+        if ("key.class".equals(property)
+            && !checkEncodingAndDecoding(coder, reader.getCurrentKey())) {
+          this.setExpectedKeyClass(reader.getCurrentKey().getClass());
```

Comment:
Can you explain exactly how you're preventing the null pointer exceptions with this change? It doesn't seem like you've fixed the root cause. I suspect all you've really done with this change is that by moving the call to getCurrentKey inside of the corner case of the validateClass, it gets called less frequently, so you're not seeing it in the scenario you were before. 

1) There's still a lot of possibility for null pointer exceptions - you do lots of things in the pattern: 
reader.getCurrentKey().getClass()
if getCurrentKey() returns null, you'll get a null pointer exception. 

2) However, looking at the original exception: 
	at org.apache.cassandra.hadoop.cql3.CqlRecordReader.getCurrentKey(CqlRecordReader.java:177)
	at org.apache.cassandra.hadoop.cql3.CqlRecordReader.getCurrentKey(CqlRecordReader.java:79)
	at org.apache.beam.sdk.io.hadoop.inputformat.HadoopInputFormatIO$HadoopInputFormatBoundedSource.validateUserInputForKeyAndValue(HadoopInputFormatIO.java:540)

the exception you're hitting was happening inside of the getCurrentKey. 

Looking at http://grepcode.com/file/repo1.maven.org/maven2/org.apache.cassandra/cassandra-all/2.2.0/org/apache/cassandra/hadoop/cql3/CqlRecordReader.java/#173 - it appears that it's a trivial method that just reads currentRow. So the cause is elsewhere, where currentRow is somehow ending up null.

basically, looking at https://github.com/diptikul/incubator-beam/blob/HIFIO-CS-ES/sdks/java/io/hadoop/input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java#L658 it appears that the problem is that fetchFirstRecordReader calls nextKeyValue, but never checks the return value.

In the case where it returns false - http://grepcode.com/file/repo1.maven.org/maven2/org.apache.cassandra/cassandra-all/2.2.0/org/apache/cassandra/hadoop/cql3/CqlRecordReader.java/#195 it is indicating that there is no data. But this code ignores that case, and tries to read getCurrentKey, and then throws a null pointer exception.